src/python/pants/engine/rules.py

```diff
@@ -230,15 +191,22 @@ def _get_starting_indent(source):
   return 0
 
 
-def _make_rule(output_type, input_selectors, for_goal=None, cacheable=True):
+def _make_rule(output_type, input_selectors, cacheable=True):
   """A @decorator that declares that a particular static function may be used as a TaskRule.
 
+  As a special case, if the output_type is a subclass of `Goal`, the `Goal.Options` for the `Goal`
+  are registered as dependency Optionables.
+
   :param type output_type: The return/output type for the Rule. This must be a concrete Python type.
   :param list input_selectors: A list of Selector instances that matches the number of arguments
     to the @decorated function.
-  :param str for_goal: If this is a @console_rule, which goal string it's called for.
   """
 
+  is_goal_cls = isinstance(output_type, type) and issubclass(output_type, Goal)
+  if is_goal_cls == cacheable:
```

Comment:
Can we remove the `cacheable` arg, and define cacheability exactly based on whether a `Goal` is returned, rather than having this check?

Reply:
Yes, but I think that that would make `@console_rule` unnecessary, and that feels like something we should wait until #6598 to nail down.